8178368: Right alignment of text fields and alignment of prompt text works incorrectly

karthikpandelu opened the pull request:
When Text width was more than TextField width, the logic to update `textTranslateX` in `updateCaretOff` method was causing the issue of unexpected behavior for Right and Center alignment.

Made changes to update `textTranslateX` in `updateCaretOff` method only when text width is less than text field width i.e `delta` is positive. 
For both right and center alignments, the `textTranslateX` value calculated in `updateTextPos` method will be updated without any condition so that expected behavior is achieved for all scenarios of text width relative to text field width. 

Added unit tests to validate LEFT, CENTER and RIGHT alignments. RIGHT and CENTER alignment tests are expected to fail without the fix provided in this PR.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8178368](https://bugs.openjdk.org/browse/JDK-8178368): Right alignment of text fields and alignment of prompt text works incorrectly


### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - Committer)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/980/head:pull/980` \
`$ git checkout pull/980`

Update a local copy of the PR: \
`$ git checkout pull/980` \
`$ git pull https://git.openjdk.org/jfx pull/980/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 980`

View PR using the GUI difftool: \
`$ git pr show -t 980`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/980.diff">https://git.openjdk.org/jfx/pull/980.diff</a>

</details>

andy-goryachev-oracle commented:
While the code change looks acceptable, I would like to point your attention to the difference in the behavior relative to the 'prompt' text.

When the prompt text length exceeds the width of the text field, it looks like the prompt text is right-aligned, and the end of the prompt text is visible (the prompt text is "<prompt>":

<img width="536" alt="Screen Shot 2022-12-21 at 15 09 19" src="https://user-images.githubusercontent.com/107069028/209020711-de1c2c27-6ee2-499f-94b0-581d289b5e5a.png">

However, setting a long text results in the beginning of the said text showing (the text string is `"<beg-01234567890123456789012345678901234567890123456789012345678901234567890123456789-end>"`):

<img width="536" alt="Screen Shot 2022-12-21 at 15 09 30" src="https://user-images.githubusercontent.com/107069028/209020786-3d4b9177-d6b7-44eb-b3a7-f8869881069e.png">

What I was expecting is the same behavior as the prompt text, i.e.:

<img width="536" alt="Screen Shot 2022-12-21 at 15 09 47" src="https://user-images.githubusercontent.com/107069028/209020871-fc013a1f-a4a7-4f7a-a7f4-3812837f2f6d.png">

You can try it out with this application -
https://github.com/andy-goryachev-oracle/Test/blob/main/src/goryachev/ensemble/TestEnsembleApp.java

andy-goryachev-oracle commented:
So the question is - do we want to change the way text is initially aligned to match the prompt text, or not?

karthikpandelu commented:
Thanks for the above details Andy.
I saw this behavior with the source code given in the bug as well.

> When the prompt text length exceeds the width of the text field, it looks like the prompt text is right-aligned, and the end of the prompt text is visible (the prompt text is "":
> 
I believe in this case, it gets aligned to the right because text width is less than text field width.
 
> However, setting a long text results in the beginning of the said text showing (the text string is `"<beg-01234567890123456789012345678901234567890123456789012345678901234567890123456789-end>"`):
> 
Yes setting a long text is resulting in this behavior. I fixed the bug based on this behavior itself.

>So the question is - do we want to change the way text is initially aligned to match the prompt text, or not?
>
Yes this have to be discussed if we want to change the initial behavior of the text alignment.

aghaisas commented:
The proposed fix fixes the reported issue.

**Regarding the question of changing current behavior of RIGHT alignment of the TextField:**
If the string is larger than the TextField width -
- LEFT align should result in - Start of the text visible and end of the text hidden.
- RIGHT align should result in - End of the text visible at the right edge of the TextField and start of the text hidden.

Right now, only the LEFT align behaves correctly.
RIGHT align behaves correctly only if the string is smaller than the TextField width. The moment string length is larger than TextField width, it starts behaving opposite (start of the text visible and end of the text hidden.)
I don't know whether there is some rationale for this implementation or it is a bug which went unnoticed till now.

andy-goryachev-oracle commented:
Actually, it looks like this use case of setting a long text on TextField with alignment *other than* LEFT is broken.  Here is it for TOP_CENTER which also is incorrect:

<img width="1112" alt="Screen Shot 2022-12-22 at 10 55 11" src="https://user-images.githubusercontent.com/107069028/209206893-f792abaa-ba2c-4802-89fa-fd2640488b8c.png">

I might recommend using the tester to enumerate all the cases where the initial text position is incorrect and fix it as a part of this ticket.  We could also modify the test to cycle through all the possible values of `Pos.values()`.

What do you think?

andy-goryachev-oracle reviewed (changes requested):
I think it's a good opportunity to test and fix all other `Pos.values()`

karthikpandelu commented:
I checked all the alignment and text width combinations using the tester. I see that prompt text is working as expected for all alignment options and the text is not.
I will fix the issues with text alignment as part of this bug.

kevinrushforth commented:
> I will fix the issues with text alignment as part of this bug.

In which case, I recommend to update the title of the JBS bug and this PR, since it is more than just right alignment that will be fixed.

karthikpandelu commented:
Sure Kevin, I will update that

karthikpandelu commented:
Fixed the issues with RIGHT and CENTER alignment of text field. Updated the details in bug and PR. Please review

kevinrushforth commented:
/reviewers 2

andy-goryachev-oracle reviewed (approved):
tested with the monkey tester

https://github.com/andy-goryachev-oracle/Test/blob/main/src/goryachev/monkey/MonkeyTesterApp.java

works beautifully with all the alignment settings.  good job, Karthik!

aghaisas reviewed (changes requested):
What I have observed is - this fixes the alignment issues, but breaks the cursor movement in case of RIGHT and CENTER alignment.

Tested using MonkeyTester application provided by Andy.
- For a Textfield, select long text and alignment to RIGHT
- In the TextField, click after text "end>" and enter text such that the string becomes longer than that the TextField can display
- Now, start pressing Left arrow key - so that the text cursor starts moving left - if we continue this key press till beginning of the string - it is expected that beginning of the text string is visible. Cursor never reaches beginning of the string with changes done in this PR.

Similar cursor movement issue is observed with CENTER alignment as well.

andy-goryachev-oracle commented:
You are right @aghaisas !
Small correction: the cursor operates correctly, the TextField just does not scroll the text so the caret remains visible.  Neither keyboard nor mouse operation (i.e. drag to select) auto-scroll the text.

For the mouse case:
- select long text
- select right or center alignment
- make the text field narrow
- click on the right end of text
- drag the mouse to the left to select
expected behavior is the text is scrolled to keep the caret always visible.

in the case of center alignment, the text does scroll when dragging to the right, but not to the left.

andy-goryachev-oracle reviewed (changes requested):
see above

hjohn commented:
> The proposed fix fixes the reported issue.
> 
> **Regarding the question of changing current behavior of RIGHT alignment of the TextField:** If the string is larger than the TextField width -
> 
> * LEFT align should result in - Start of the text visible and end of the text hidden.
> * RIGHT align should result in - End of the text visible at the right edge of the TextField and start of the text hidden.
> 
> Right now, only the LEFT align behaves correctly. RIGHT align behaves correctly only if the string is smaller than the TextField width. The moment string length is larger than TextField width, it starts behaving opposite (start of the text visible and end of the text hidden.) I don't know whether there is some rationale for this implementation or it is a bug which went unnoticed till now.

I'm not entirely convinced this is a bug -- it seems quite intentionally built this way.  

A reason why you may want to show the left hand side of the text if it doesn't fit is that the exact length of text and the exact size of a (resizable) UI is hard to predict before hand.  When presented with a new control where the text doesn't fit, it may make more sense and be more natural to show the user the start of the text.

We may want to check how other UI toolkits are handling these cases.

In any case, this current fix has a problem where the cursor can get lost (field is not scrolled to show it), while the current implementation has no such issues.

karthikpandelu commented:
I checked with HTML textfield and swing JTextField. In both the cases, the text gets aligned to the left (beginning of the text is shown) regardless of alignment type when text length is more than text field length.
In html textfield, even prompt text has same behavior but in JavaFX prompt text gets aligned to the center (center of the text is displayed) or right (beginning of the text is hidden) as specified even when the prompt text length is more than textfield length.

The case where prompt text length is more than text field length might not be frequent but I think we should have same behavior for both. So I think we should fix either the behavior of the text or the  prompt text so that both are consistent.

I went ahead with the option of fixing the behavior of text because aligning the text according to the set value regardless of the text length looked like predictable behavior.
Let me know your thoughts on this.

andy-goryachev-oracle commented:
I think the behavior with the fix is correct - it should follow the alignment setting.

@hjohn is right on one account - scroll to cursor is not working right.  to reproduce
- set the alignment to either *_RIGHT or *_CENTER
- set a long text string (longer than can fit the control)
- click on the visible area and repeatedly move cursor left beyond the left edge.
Expected:
-- the view should scroll horizontally to bring the caret back into the view
Observed with the fix:
-- it doesn't

kevinrushforth commented:
In this case I think we should match the behavior of JTextField and html, where the beginning of the text is always shown. It seems more consistent and less surprising.

The behavior prompt text should match whatever we do for text.

andy-goryachev-oracle commented:
So, regardless of the alignment setting, whether the width of the control is less than the space required to show all the text (or all of the prompt text), we should show the beginning of the text.  Makes sense!
And, for RTL orientation, that will be mirrored - but still shows the beginning (of RTL) text.

karthikpandelu commented:
Sure. I'll revert the changes made for right and center alignment and make the prompt text alignment consistent with the text behavior.
I will update the bug and PR accordingly.

karthikpandelu commented:
Fixed the text alignment issue and changed the prompt text alignment behavior in consistent with text alignment behavior.
The text and prompt text alignment behavior will be as follows:

- When text/prompt length is less than text field width, text/prompt aligns according to the alignment type given.

- When text/prompt length is more than text field width, text/prompt aligns to the left showing the start of the text regardless of the alignment type given. 

Please review these changes.

hjohn commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextFieldSkin.java:
```diff
@@ -802,25 +802,41 @@ private void updateTextPos() {
-            // Update if there is space on the right
-            if (newX + textNodeWidth <= textRight.get()) {
-                textTranslateX.set(newX);
+                // Update if there is space on the right
+                if (newX + textNodeWidth <= textRight.get()) {
+                    textTranslateX.set(newX);
+                } else if (newX < 0 && oldX > 1) {
+                    textTranslateX.set(caretWidth / 2);
+                }
```
I get the impression this code also needs to use `caretWidth / 2` instead of `0` and `1`.

ie:

```suggestion
                // Update if there is space on the right
                if (newX + textNodeWidth <= textRight.get() - caretWidth / 2) {
                    textTranslateX.set(newX);
                } else if (newX < caretWidth / 2 && oldX > caretWidth / 2) {
                    textTranslateX.set(caretWidth / 2);
                }
```

It would only work for very narrow caret's otherwise, and for wide caret's it would have some unexpected positioning in the edge case where text almost fits.

karthikpandelu replied:
> I get the impression this code also needs to use `caretWidth / 2` instead of `0` and `1`.
> 
Here also same comment as above for the `newX < caretWidth/2` condition. But while comparing the `oldX`, `caretWidth/2` should be considered instead of 1 and also `caretWidth / 2` is subtracted from `textRight` value in the previous if condition as you suggested. These updates have been done in the code.

> I would also recommend testing the code on Windows with the screen scale set to 225%, as it might show issues related to fractional scale.

I will test this latest code updates.

karthikpandelu replied:
Tested in windows with screen scale set to as high as 225%. Couldn't reproduce any issues with updated code.

andy-goryachev-oracle replied:
thank you for checking!

hjohn commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextFieldSkin.java:
```diff
@@ -802,25 +802,41 @@ private void updateTextPos() {
+            } else {
+                newX = textRight.get() - textNodeWidth - caretWidth / 2;
+                // Update if there is space on the right
+                if (newX > oldX || newX > 0) {
+                  textTranslateX.set(newX);
+                } else if (newX < 0 && oldX > 1) {
+                  textTranslateX.set(caretWidth / 2);
+                }
```
Again, I don't think its correct.  It's just odd that `newX` may be smaller than `caretWidth / 2` but is accepted, but when the text is too wide the "minimum" value suddenly is `caretWidth / 2`.

karthikpandelu replied:
Same comment as above for `newX` comparison with `caretWidth/2`. Updated `caretWidth/2` instead of 1 while comparing `oldX` value.

hjohn commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextFieldSkin.java:
```diff
@@ -802,25 +802,41 @@ private void updateTextPos() {
-                                      caretWidth / 2);
+                double promptOldX = promptNode.getLayoutX();
+                double promptNewX = textRight.get() - promptNode.getLayoutBounds().getWidth() - caretWidth / 2;
+                if(promptNewX > promptOldX || promptNewX > 0) {
+                    promptNode.setLayoutX(promptNewX);
+                } else {
+                    // Align to left when prompt text length is more than text field width
+                    promptNode.setLayoutX(caretWidth / 2);
+                }
```
Similar comment, I don't think its correct. It's just odd that promptNewX  may be smaller than caretWidth / 2 but is accepted, but when the text is too wide the "minimum" value suddently is caretWidth / 2.

Also, I don't see how `promptOldX` has any relevance when it comes to positioning the prompt.  I think this is only relevant for the actual text (because it can scroll depending on where the cursor is), not for the prompt -- unless the prompt can be scrolled as well.  If that's the case then the `CENTER` case may need to be updated to take `promptOldX` into account as well.  As it stands currently, they have behaviors that seem to conflict.

karthikpandelu replied:
Yes `promptOldX` is not required in case of prompt text. Code is updated.
Same comment as the first one for the condition while comparing newly calculated value with `careWidth/2`.

hjohn commented:
I think this class may benefit from a few tests that test with a very wide caret, to see if positioning is what you'd expect in those cases as well.  I get the impression a lot of the code assumes a narrow caret (1 or 2 pixels) and this is why we see constants like `0` and `1` in the code, and even places where the caret width should be subtracted but isn't because it is assumed to be `0`.

andy-goryachev-oracle commented:
Noticed an issue similar to what @aghaisas reported earlier.
- set TOP_RIGHT alignment (any _RIGHT in fact)
- resize the field such that the text is wider than TextField can display
- click on the first visible character and start typing
Expected: the typed chars should appear in the view, the caret moves right, the remaining text scrolls right
Observed: the typed characters appear to the left of the caret which does not move.

It might behave differently when the caret is at the rightmost position next to the control edge, in which case the behavior is correct.  Perhaps there ought to be some conditional logic implemented, but the way it behaves now when the caret is close to the left edge feels unnatural.

![Screenshot 2023-02-23 at 09 32 55](https://user-images.githubusercontent.com/107069028/220986606-1990cc60-1d33-4894-b93f-c9eac2202908.png)

andy-goryachev-oracle reviewed (commented):
almost there, except for one scenario with _RIGHT alignment which feels weird.

karthikpandelu commented on modules/javafx.controls/src/main/java/javafx/scene/control/skin/TextFieldSkin.java:
```diff
@@ -802,25 +802,41 @@ private void updateTextPos() {
                 // appear at the left of the centered prompt.
                 newX = midPoint - promptNode.getLayoutBounds().getWidth() / 2;
-                promptNode.setLayoutX(newX);
+                if (newX > 0) {
```
Thanks for these details @andy-goryachev-oracle . I was trying to figure out how I can change the caret width. Just to check the behavior of text alignment, I changed the hard coded value at line no 233.

> Let's say `caretWidth / 2` is 5, and that would be position we would show the prompt text if it doesn't fit.
> 
> If the `newX` calculation is less than 5, but greater than 0, the prompt text would be further to the left than you'd expect.
> 
> So I think this line should be:
> 
> Probably best to put that in a local variable.

The newX is calculated symmetrically from midpoint using the text length. If we consider the condition `newX > careWidth/2`, we will be aligning the text at the position of `caretWidth/2` even when there is space for the whole text. Which means that the text will be aligned to left even when there is space for whole text with CENTER alignment. So in my opinion, current conditions and values gives better alignment behavior.

hjohn replied:
> Thanks for these details @andy-goryachev-oracle . I was trying to figure out how I can change the caret width. Just to check the behavior of text alignment, I changed the hard coded value at line no 233.
> 
> > Let's say `caretWidth / 2` is 5, and that would be position we would show the prompt text if it doesn't fit.
> > If the `newX` calculation is less than 5, but greater than 0, the prompt text would be further to the left than you'd expect.
> > So I think this line should be:
> > Probably best to put that in a local variable.
> 
> The newX is calculated symmetrically from midpoint using the text length. If we consider the condition `newX > careWidth/2`, we will be aligning the text at the position of `caretWidth/2` even when there is space for the whole text. Which means that the text will be aligned to left even when there is space for whole text with CENTER alignment. So in my opinion, current conditions and values gives better alignment behavior.

Alright, thanks for checking this.

karthikpandelu commented:
> I think this class may benefit from a few tests that test with a very wide caret, to see if positioning is what you'd expect in those cases as well. I get the impression a lot of the code assumes a narrow caret (1 or 2 pixels) and this is why we see constants like `0` and `1` in the code, and even places where the caret width should be subtracted but isn't because it is assumed to be `0`.

Since caret width is hard coded and there is no direct way of changing the width, I believe we can proceed without adding tests with wide caret.

karthikpandelu commented:
> It might behave differently when the caret is at the rightmost position next to the control edge, in which case the behavior is correct. Perhaps there ought to be some conditional logic implemented, but the way it behaves now when the caret is close to the left edge feels unnatural.
> 
Yes this behavior looks unnatural. I see this issue in the mainline as well.
Basically this behavior is observed because of the logic that, when we start typing in text field with RIGHT alignment, left side of the text is scrolled. So from whichever position we start inserting new characters, left side of the text gets scrolled. Because of the same behavior the newly inserted text scrolls to the left when we start typing from the first visible character. I will analyze a bit more on this behavior.

hjohn commented:
> > I think this class may benefit from a few tests that test with a very wide caret, to see if positioning is what you'd expect in those cases as well. I get the impression a lot of the code assumes a narrow caret (1 or 2 pixels) and this is why we see constants like `0` and `1` in the code, and even places where the caret width should be subtracted but isn't because it is assumed to be `0`.
> 
> Since caret width is hard coded and there is no direct way of changing the width, I believe we can proceed without adding tests with wide caret.

Yeah, that's okay; it seems to make a lot of assumptions about the caret, even going as far as ignoring a caret with exactly 4 path elements (as apparently there is no way to ask if it is a split caret when RTL/LTR mixed text is encountered).  I guess those will need to tackled if we ever change the caret shape.

andy-goryachev-oracle commented:
> Yes this behavior looks unnatural. I see this issue in the mainline as well.

I think the condition to select one behavior or the other might be whether all the text fits in the visible area.  If it does - use existing logic.  If it does not - shift the text to the right.

Question: should we fix this as part of this PR (is it easy enough?) or does it warrant a separate Bug/RFE?

karthikpandelu commented:
> I think the condition to select one behavior or the other might be whether all the text fits in the visible area. If it does - use existing logic. If it does not - shift the text to the right.

Yes this approach can be used. But I think there is no need to change anything in `updateCaretOff` or `updateTextPos` methods.
This behavior exists only in RIGHT alignment because of the code present in `scrollAfterDelete` method. In this method, LEFT and CENTER alignments cases have same block of code and only RIGHT alignment has different case statement. Just to check the behavior, I applied LEFT and CENTER alignment code  to RIGHT alignment as well in this method and observed that the RIGHT alignment behavior changed to the expected one without any side effects. But I'm yet to figure out why exactly there is separate case for RIGHT alignment.

> 
> Question: should we fix this as part of this PR (is it easy enough?) or does it warrant a separate Bug/RFE?

I think once we figure out why exactly there is separate case for RIGHT alignment in `scrollAfterDelete` method, we can decide if this can be fixed in this PR or we should create separate bug.

Also I did quick check in HTML and JTextField. In both, text scrolls to the right when we click on the first visible character and start typing.

andy-goryachev-oracle commented:
thank you for checking!

karthikpandelu commented:
Fixed the issue of text not getting scrolled on typing from left edge of the right aligned text whose length is more than the text field length.
The switch cases present in `scrollAfterDelete()` is not required as the block of code which is present for case RIGHT induces the unwanted behavior. Conditions to update `textTranslateX` in this method can be same for all the type of alignments.

andy-goryachev-oracle reviewed (approved):
TextField behavior looks good in the tester!  The monkey says thank you.

karthikpandelu commented:
Resolved the merge conflict. Please review.

kevinrushforth reviewed (approved):
Looks good now.

karthikpandelu commented:
/integrate

aghaisas commented:
/sponsor
